src/utils/provider.ts

```diff
@@ -0,0 +1,37 @@
+import { useContext, useState } from 'react'
+import { defaultConfig } from './config'
+import { wrapCache } from './cache'
+import { isFunction, UNDEFINED } from './helper'
+import { SWRConfigContext } from './config-context'
+import { SWRGlobalState, GlobalState } from './global-state'
+
+import { Cache, ProviderOptions } from '../types'
+
+export const useSWRProvider = <Data = any>(
+  initialize?:
+    | ((currentCache: Readonly<Cache>) => Cache)
+    | Partial<ProviderOptions>,
+  options?: Partial<ProviderOptions>
+) => {
+  const cache = useContext(SWRConfigContext).cache || defaultConfig.cache
+
+  // We use a state to keep the cache instance and make sure it's only
+  // initialized once.
+  const [newProvider] = useState(() => {
```

Comment:
Consider adding `cache` only to SWRConfig, and initialize/wrap cache inside SWRConfig.

```js
<SWRConfig value={{ cache: () => (new Map()) }}>
  <Page>
</SWRConfig>
```

For consuming side only use `{ mutate, cache } = useSWRProvider()`. then we could deprecate global `mutate`.

Reply:
It's possible to deprecate `mutate` in the future, but still early to say. The simplicity of the current API is great I feel if you don't customize the cache or use any concurrent rendering techniques.

Reply:
> Consider adding cache only to SWRConfig, and initialize/wrap cache inside SWRConfig.

This is a good idea, I need to validate it.

Reply:
The good part of this approach is that we can avoid the problem of using `useSWRProvider` to create new cache at a component level:

```js
function useUser() {
  const { cache } = useSWRProvider(() => new Map) // this should be avoided
  return useSWR('/api/user', { cache })
}
```

If we only allow creating the provider within `<SWRConfig>`, then it will be at the boundary level.

As a potential downside, it's tricky to pass `cache` to the SWR hook directly (_Is this still a valid case?_). Maybe we need `createCache` for it, but it is definitely not concurrent safe. I'd leave this use case for future versions.